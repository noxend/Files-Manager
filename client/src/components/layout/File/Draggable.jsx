import React from 'react';
import { Icon } from 'antd';

export default View =>
  class extends React.Component {
    state = {
      startPosition: { x: 0, y: 0 },
      mouseStartPosition: { x: 0, y: 0 }
    };

    componentWillMount = () => {
      this.dragGhost = React.createRef();
    };

    setPosition = (x, y) => {
      this.dragGhost.current.style.transform = `translate(${x}px, ${y}px)`;
    };

    onDragEnd = e => {
      e.currentTarget.classList.toggle('--drag');
      this.setPosition(this.state.startPosition.x, this.state.startPosition.y);
      this.dragGhost.current.style.opacity = 0;
      setTimeout(() => {
        this.dragGhost.current.classList.toggle('--move');
      }, 300);
    };

    onDragStart = e => {
      const { fileId } = this.props;
      const item = e.currentTarget.cloneNode();
      const { left, top } = e.currentTarget.getBoundingClientRect();
      e.dataTransfer.setDragImage(item, 0, 0);
      e.dataTransfer.setData('fileId', fileId);
      e.currentTarget.classList.toggle('--drag');
      setTimeout(() => {
        this.dragGhost.current.classList.toggle('--move');
      }, 0);
      this.dragGhost.current.style.opacity = 1;
      this.setState({
        startPosition: {
          x: e.currentTarget.offsetLeft,
          y: e.currentTarget.offsetTop
        },
        mouseStartPosition: {
          x: left,
          y: top
        }
      });
    };

    onDrag = e => {
      const { x, y } = this.state.mouseStartPosition;
      if (
        e.clientX + this.dragGhost.current.offsetWidth + 5>=
        document.body.offsetWidth
      ) {
        this.setPosition(
          e.clientX - x - this.dragGhost.current.offsetWidth - 5,
          e.clientY - y + 5
        );
      } else {
        this.setPosition(e.clientX - x + 5, e.clientY - y + 5);
      }
    };

    render = () => {
      return (
        <div style={{ position: 'relative' }}>
          <View {...this.props} {...this} />
          <div ref={this.dragGhost} className="drag-ghost">
            <Icon
              type="file"
              style={{ fontSize: '18px', marginRight: '8px' }}
            />
            <span>{this.props.fileName}</span>
          </div>
        </div>
      );
    };
  };
