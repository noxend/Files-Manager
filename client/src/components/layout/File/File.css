.file {
  height: 195px;
  display: flex;
  flex-direction: column;
  background-color: white;
  border-radius: 4px;
  overflow: hidden;
  border-radius: 4px;
  border: 1px solid rgb(235, 237, 240);
  cursor: pointer;
  transition: 0.2s;
  box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
  opacity: 1;
  transform: scale(1);
}

.--drag {
  opacity: 0.3;
  transform: scale(0.9);
}

.file:hover {
  background-color: rgba(0, 0, 0, 0.015);
  box-shadow: 0 5px 5px rgba(0, 0, 0, 0.03);
}

.file__top {
  position: relative;
  width: 100%;
  flex-grow: 1;
  display: flex;
  justify-content: center;
  align-items: center;
  overflow: hidden;
}

.file__bottom {
  padding: 10px 14px;
  display: flex;
  justify-content: space-between;
}

.file__color {
  background: linear-gradient(45deg, #f12711, #f5af19);
  width: 100%;
  height: 100%;
  position: absolute;
  opacity: 0.6;
}

.file__bg {
  position: absolute;
  width: 100%;
  height: 100%;
  filter: blur(15px);
  transform: scale(2.3);
  background-size: 200%;
  opacity: 0.65;
}

.file__icon {
  position: absolute;
  height: 120px;
  width: 120px;
  filter: drop-shadow(5px 5px 0 rgba(0, 0, 0, 0.07));
}

.file__name {
  font-weight: 500;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.file__size {
  white-space: nowrap;
  margin-left: 10px;
}

.file-skeleton {
  height: 195px;
  border-radius: 4px;
  background: linear-gradient(90deg, #f2f2f2 25%, #e6e6e6 37%, #f2f2f2 63%);
  background-size: 500% 100%;
  animation: skeleton-loading 1.4s ease infinite;
}

.drag-ghost {
  left: 0;
  top: 0;
  z-index: -20;
  position: absolute;
  max-width: 220px;
  background-color: rgb(251, 251, 251);
  border: 1px solid rgb(235, 237, 240);
  padding: 10px 14px;
  box-shadow: 0 10px 20px rgba(0, 0, 0, 0.05), 0 6px 6px rgba(0, 0, 0, 0.03);
  border-radius: 4px;
  display: none;
  user-select: none;
  /* justify-content: center; */
  align-items: center;
  opacity: 0;
}

.drag-ghost span {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  font-weight: 500;
}

.--move {
  z-index: 20;
  display: flex;
}

@keyframes skeleton-loading {
  0% {
    background-position: 100% 50%;
  }
  100% {
    background-position: 0 50%;
  }
}
