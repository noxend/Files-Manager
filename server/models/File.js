const knex = require('../knex');
const { TransformFields, createModel } = require('../libs/model');

class FileModel extends TransformFields {
  constructor({ name, originalName, size }) {
    super();
    this.fields = {
      original_name: originalName,
      file_name: name,
      file_size: size
    };
  }

  static async getUserFiles(folderId) {
    const result = await knex
      .select(
        'files.id as fileId',
        'original_name as fileName',
        'file_size as fileSize',
        'files.created_at as createdAt'
      )
      .from('users_files')
      .join('files', 'files.id', 'users_files.file_id')
      .where('users_files.folder_id', '=', folderId);

    return result;
  }
}

module.exports = createModel('files', FileModel);
