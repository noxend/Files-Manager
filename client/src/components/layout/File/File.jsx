import React from 'react';
import CustomIcon from '../Icons/CustomIcon';
import Draggable from './Draggable';

import './File.css';

const File = props => {
  const {
    type,
    fileName,
    fileSize,
    onClick,
    onDragEnd,
    onDragStart,
    onDrag
  } = props;

  const formatFileSize = (bytes, dm = 2) => {
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(1024));
    return `${(bytes / Math.pow(1024, i)).toFixed(dm)} ${sizes[i]}`;
  };

  return (
    <div
      className="file"
      onClick={onClick}
      draggable={true}
      onDragStart={onDragStart}
      onDragEnd={onDragEnd}
      onDrag={onDrag}
    >
      <div className="file__top">
        <div className="file__color" />
        <CustomIcon className="file__icon" type={type} />
      </div>
      <div className="file__bottom">
        <span className="file__name">{fileName}</span>
        <span className="file__size">{formatFileSize(fileSize)}</span>
      </div>
    </div>
  );
};

export default Draggable(File);
