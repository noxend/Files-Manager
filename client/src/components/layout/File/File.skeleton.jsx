import React from 'react';

import './File.css';

const File = props => {

  return <div className="file-skeleton" />;
};

export default File;
