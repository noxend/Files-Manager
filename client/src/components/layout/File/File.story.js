import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, select, text } from '@storybook/addon-knobs';

import File from './File';

const stories = storiesOf('Buttons', module);

stories.addDecorator(withKnobs);

stories.add('File Button', () => {
  const type = select(
    'Types',
    { PSD: 'psd', DOCX: 'docx', JPEG: 'jpg' },
    'psd'
  );

  const fileName = text('File Name', 'filename');

  return (
    <div style={{ width: '260px' }}>
      <File type={type} fileName={fileName} />
    </div>
  );
});
