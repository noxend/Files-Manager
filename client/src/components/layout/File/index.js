import File from './File';
import FileSkeleton from './File.skeleton';

export { File, FileSkeleton };
