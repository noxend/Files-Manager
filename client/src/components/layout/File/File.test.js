import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import File from './File';

configure({ adapter: new Adapter() });

describe('<File />', () => {
  it('Snapshot', () => {
    const tree = renderer
      .create(<File type="psd" fileName="v2.psd" fileSize={1073741824} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Props should to match the component', () => {
    const wrapper = shallow(<File type="psd" fileName="v2.psd" fileSize={1073741824} />);
    expect(wrapper.find('.file__name').text()).toBe('v2.psd');
    expect(wrapper.find('.file__icon').prop('type')).toBe('psd');
    expect(wrapper.find('.file__size').text()).toBe('1.00 GB');
  });
});
