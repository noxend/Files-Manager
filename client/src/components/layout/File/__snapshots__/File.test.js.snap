// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<File /> Snapshot 1`] = `
<div
  className="file"
>
  <div
    className="file__top"
  >
    <svg
      className="file__bg"
      type="psd"
    >
      psd.svg
    </svg>
    <div
      className="file__color"
    />
    <svg
      className="file__icon"
      type="psd"
    >
      psd.svg
    </svg>
  </div>
  <div
    className="file__bottom"
  >
    <span
      className="file__name"
    >
      v2.psd
    </span>
    <span
      className="file__size"
    >
      1.00 GB
    </span>
  </div>
</div>
`;
